src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/Cas1OutOfServiceBedDetailsEntity.kt

```diff
@@ -0,0 +1,74 @@
+package uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity
+
+import org.springframework.data.jpa.repository.JpaRepository
+import org.springframework.stereotype.Repository
+import java.time.LocalDate
+import java.time.OffsetDateTime
+import java.util.EnumSet
+import java.util.UUID
+import javax.persistence.Column
+import javax.persistence.Entity
+import javax.persistence.EnumType
+import javax.persistence.Enumerated
+import javax.persistence.Id
+import javax.persistence.JoinColumn
+import javax.persistence.ManyToOne
+import javax.persistence.Table
+
+@Repository
+interface Cas1OutOfServiceBedDetailsRepository : JpaRepository<Cas1OutOfServiceBedDetailsEntity, UUID>
+
+@Entity
+@Table(name = "cas1_out_of_service_bed_details")
+data class Cas1OutOfServiceBedDetailsEntity(
+  @Id
+  val id: UUID,
+  val createdAt: OffsetDateTime,
+  @Enumerated(EnumType.STRING)
+  val detailType: Cas1OutOfServiceBedDetailType,
+  var startDate: LocalDate,
+  var endDate: LocalDate,
+  var referenceNumber: String?,
+  var notes: String?,
+  @ManyToOne
+  @JoinColumn(name = "out_of_service_bed_reason_id")
+  var reason: Cas1OutOfServiceBedReasonEntity,
+  @ManyToOne
+  @JoinColumn(name = "out_of_service_bed_id")
+  val outOfServiceBed: Cas1OutOfServiceBedEntity,
+  @ManyToOne
+  @JoinColumn(name = "created_by_user_id")
+  val createdBy: UserEntity,
+  @Column(name = "change_type")
+  val changeTypePacked: Long,
+)
+
+enum class Cas1OutOfServiceBedDetailType {
+  INITIAL,
+  UPDATE,
+}
+
+@Suppress("detekt:MagicNumber")
+enum class Cas1OutOfServiceBedDetailChangeType(private val value: Long) {
+  START_DATE(0b0000_0001),
+  END_DATE(0b0000_0010),
+  REFERENCE_NUMBER(0b0000_0100),
+  REASON(0b0000_1000),
+  NOTES(0b0001_0000),
+  ;
+
+  companion object {
+    const val NO_CHANGE: Long = 0
+
+    fun pack(values: EnumSet<Cas1OutOfServiceBedDetailChangeType>): Long = values
+      .fold(0L) { acc, changeType -> acc or changeType.value }
+
+    fun unpack(value: Long): EnumSet<Cas1OutOfServiceBedDetailChangeType> {
+      val result = Cas1OutOfServiceBedDetailChangeType.entries.filter { (it.value and value) != 0L }
+
+      return EnumSet
+        .noneOf(Cas1OutOfServiceBedDetailChangeType::class.java)
+        .apply { addAll(result) }
+    }
```

Comment:
this looks really neat!

I see you've added some notes in the PR narrative:

> To provide rich information about the scope of each update, when a new Cas1OutOfServiceBedDetailsEntity is created a comparison is done against the most recent existing one for the start date, end date, reason ID, reference number, and notes, which is then packed into the change_type column as bit flags. This is unpacked into a set of Cas1OutOfServiceBedHistoryType values when marshalling the API response. This approach avoids having to perform the comparison for each history item whenever an out-of-service bed record is requested.

I wonder if future devs/selves would be grateful to have an expanded version of this note included in the commit message here? Perhaps include some online references / explanation of the binary values used?

Can we describe the behaviour we're optimising or overriding here in a unit test?